pkg/resource/deploy/deployment.go

```diff
@@ -80,6 +82,100 @@ func (o Options) InfiniteParallelism() bool {
 	return o.Parallel == math.MaxInt32
 }
 
+// An immutable set of urns to target with an operation.
+//
+// The zero value of UrnTargets is the set of all URNs.
+type UrnTargets struct {
+	// UrnTargets is internally made up of two components: literals, which are fully
+	// specified URNs and globs, which are partially specified URNs.
+
+	literals []resource.URN
+	globs    map[string]*regexp.Regexp
+}
+
+// Create a new set of targets.
+//
+// Each element is considered a glob if it contains any '*' and an URN otherwise. No other
+// URN validation is performed.
+func NewUrnTargets(urnOrGlobs []string) UrnTargets {
+	literals, globs := []resource.URN{}, map[string]*regexp.Regexp{}
+	for _, urn := range urnOrGlobs {
+		if strings.ContainsRune(urn, '*') {
+			globs[urn] = nil
+		} else {
+			literals = append(literals, resource.URN(urn))
+		}
+	}
+	return UrnTargets{literals, globs}
+}
+
+// Create a new set of targets from fully resolved URNs.
+func NewUrnTargetsFromUrns(urns []resource.URN) UrnTargets {
+	return UrnTargets{urns, nil}
+}
+
+// Return if the instance was explicitly created, or if it was default initialized.
+func (t UrnTargets) IsInitialized() bool {
```

Comment:
IsInitialized reads a bit odd, maybe IsEmpty? Because `NewUrnTargets([]string{})` is initialized but would return false from this method.

Reply:
Good point. I'm going to change it to `IsConstrained`. 

Conceptually, `NewUrnTargets([]string{"foo*", "bar"})` is `{ urn | urn in Urns and (urn = "bar" or urn in expand("foo*")) }` (urn is constrained). 

`NewUrnTargets([]string{})` is equivalent to `{ urn | urn in Urns }` (urn is unconstrained).